include/rocksdb/trace_record.h

```diff
@@ -0,0 +1,232 @@
+//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#pragma once
+
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "rocksdb/rocksdb_namespace.h"
+#include "rocksdb/slice.h"
+#include "rocksdb/write_batch.h"
+
+// To do: ROCKSDB_LITE ?
+
+namespace ROCKSDB_NAMESPACE {
+
+class ColumnFamilyHandle;
+class DB;
+class Status;
+
+// Supported trace record types.
+enum TraceType : char {
+  kTraceNone = 0,
+  kTraceBegin = 1,
+  kTraceEnd = 2,
+  // Query level tracing related trace types.
+  kTraceWrite = 3,
+  kTraceGet = 4,
+  kTraceIteratorSeek = 5,
+  kTraceIteratorSeekForPrev = 6,
+  // Block cache tracing related trace types.
+  kBlockTraceIndexBlock = 7,
+  kBlockTraceFilterBlock = 8,
+  kBlockTraceDataBlock = 9,
+  kBlockTraceUncompressionDictBlock = 10,
+  kBlockTraceRangeDeletionBlock = 11,
+  // IO tracing related trace type.
+  kIOTracer = 12,
+  // Query level tracing related trace type.
+  kTraceMultiGet = 13,
+  // All trace types should be added before kTraceMax
+  kTraceMax,
+};
+
+class WriteQueryTraceRecord;
+class GetQueryTraceRecord;
+class IteratorSeekQueryTraceRecord;
+class MultiGetQueryTraceRecord;
+
+// Base class for all types of trace records.
+class TraceRecord {
+ public:
+  TraceRecord();
+  explicit TraceRecord(uint64_t timestamp);
+  virtual ~TraceRecord();
+
+  virtual TraceType type() const = 0;
+
+  virtual uint64_t timestamp() const;
+
+  class Handler {
+   public:
+    virtual ~Handler() {}
+
+    virtual Status Handle(const WriteQueryTraceRecord& record) = 0;
+    virtual Status Handle(const GetQueryTraceRecord& record) = 0;
+    virtual Status Handle(const IteratorSeekQueryTraceRecord& record) = 0;
+    virtual Status Handle(const MultiGetQueryTraceRecord& record) = 0;
+  };
+
+  class ExecutionHandler : public Handler {
+   public:
+    ExecutionHandler(DB* db, const std::vector<ColumnFamilyHandle*>& handles);
+
+    ExecutionHandler(
+        DB* db,
+        const std::unordered_map<uint32_t, ColumnFamilyHandle*>& cf_map);
+
+    virtual ~ExecutionHandler() override;
+
+    virtual Status Handle(const WriteQueryTraceRecord& record) override;
+    virtual Status Handle(const GetQueryTraceRecord& record) override;
+    virtual Status Handle(const IteratorSeekQueryTraceRecord& record) override;
+    virtual Status Handle(const MultiGetQueryTraceRecord& record) override;
+
+   private:
+    DB* db_;
+    std::unordered_map<uint32_t, ColumnFamilyHandle*> cf_map_;
+  };
```

Comment:
Can we expose `NewExecutionHandler(DB* db, const std::vector<ColumnFamilyHandle*>& handles)` factory function only and keep the class definition internal?

Reply:
Done